src/stats/lexd.rs

```diff
@@ -0,0 +1,57 @@
+use std::collections::HashSet;
+
+use tree_sitter::{Language, Parser, TreeCursor};
+
+use rocket_contrib::{json, json::JsonValue};
+
+use slog::Logger;
+
+use crate::{models::StatKind, stats::StatsError};
+
+extern "C" {
+    fn tree_sitter_lexd() -> Language;
+}
+
+pub fn get_stats(_logger: &Logger, body: &str) -> Result<Vec<(StatKind, JsonValue)>, StatsError> {
+    let mut parser = Parser::new();
+    let language = unsafe { tree_sitter_lexd() };
+    parser.set_language(language).unwrap();
+    let tree = parser.parse(body, None).unwrap();
+    let mut lexicons: HashSet<String> = HashSet::new();
+    let mut patterns: HashSet<String> = HashSet::new();
+    let mut lex_entries: usize = 0;
+    let mut pat_entries: usize = 0;
+
+    let mut walker: TreeCursor = tree.root_node().walk();
+
+    for child in tree.root_node().children(&mut walker) {
+        let mut walker2: TreeCursor = child.walk();
+        if child.kind() == "pattern_block" {
+            if child.child(0).unwrap().kind() == "pattern_start" {
+                patterns.insert("".to_string());
+            }
+            for line in child.children(&mut walker2) {
+                if line.kind() == "pattern_line" {
+                    pat_entries += 1;
+                } else if line.kind() == "identifier" {
+                    patterns.insert(body[line.byte_range()].to_string());
```

Comment:
Hmm, do you really need to use owned strings here? You should be able to take references into the existing string (declare the `Hash<&str>`). That should avoid a _ton_ of allocs.